apisix/plugins/auth-casdoor.lua

```diff
@@ -0,0 +1,126 @@
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local core = require("apisix.core")
+local cjson = require("cjson")
+local http = require("resty.http")
+local session = require("resty.session")
+local log = core.log
+
+local plugin_name = "auth-casdoor"
+local schema = {
+    type = "object",
+    properties = {
+        endpoint_addr = {type = "string", pattern = "^[^%?]+[^/]$"},
+        client_id = {type = "string"},
+        client_secret = {type = "string"},
+        callback_url = {type = "string", pattern = "^[^%?]+[^/]$"},
+        test = {type = "boolean", default = false}
+    },
+    required = {
+        "callback_url", "endpoint_addr", "client_id", "client_secret"
+    }
+}
+
+local _M = {
+    version = 0.1,
+    priority = 2559,
+    name = plugin_name,
+    schema = schema
+}
+
+local function fetch_access_token(ctx, conf)
+    local args = core.request.get_uri_args(ctx)
+    if not args or not args.code or not args.state then
+        return nil, "failed when accessing token. Invalid cofde or state"
+    end
+    local client = http.new()
+    local url = conf.endpoint_addr .. "/api/login/oauth/access_token"
+
+    if conf.test then
+        url = conf.endpoint_addr .. "/casdoor_fake_access_token_api"
+    end
+
+    local res, err = client:request_uri(url, {
+        method = "POST",
+        query = {
+            code = args.code,
+            grant_type = "authorization_code",
+            client_id = conf.client_id,
+            client_secret = conf.client_secret
+        }
+    })
+    if not res then return nil, err end
+    local data, err = cjson.decode(res.body)
+
+    if err or not data then return nil, err end
+
+    if not data.access_token then
+        return nil, "failed when accessing token: no access_token contained"
+    end
+
+    return data.access_token, nil
+end
+
+function _M.check_schema(conf) return core.schema.check(schema, conf) end
+
+function _M.access(conf, ctx)
+    -- log.info("hit auth-casdoor access")
+    local current_uri = ctx.var.uri
+    local session_obj_read, session_present = session.open()
+
+    -- step 1: check whether hits the callback
+    if current_uri == conf.callback_url:match(".-//[^/]+(/.*)") then
+        local access_token, err = fetch_access_token(ctx, conf)
+        if access_token then
+            if not session.present then
```

Comment:
```suggestion
            if not session_present  then
```
and need a test to cover it